design_docs/rsyslog_new_input_output_roles_proposal.md

```diff
@@ -0,0 +1,254 @@
+Considering the inputs, the functionality that `rsyslog_default` offers has to be achieved by the logging role itself instead of copying the hardcoded rsyslog.conf to the target host.
+
+This proposal will discuss about these items:
+- how to replace the `rsyslog_default` variable with the logging role input/output role configuration.
+- how to support more generic way to connect input roles and output roles.
+
+Please note that this github commit starting this doc implements the first cut of the proposal including:
+- input roles - basics input role (`imjournal`, `imuxsock`, `imudp`, `imptcp`) and files input role (`imfile`), and
+- output roles - files output role (`omfile`) and forwards output role (`omfwd`).
+Leftover tasks, other roles to be changed and eliminating `rsyslog_default`, are discussed in [To Do's](#to-dos).
+
+## Proposals
+
+### Output roles
+Each output role would be stored in a configuration file which suffix is a type name.
+
+Files and Forwards output configuration files in /etc/rsyslog.d:
+```
+  10-output-modules.conf
+  30-output.files
+  30-output.forwards
+```
+
+### Ruleset to switch the output
+Using rsyslog rulesets, we prepare all possible output combinations in the ruleset configuration file as follows.  
+Note: The ruleset name is made from the output types in the alphabetical order.
+```
+==> 50-default-rulesets.conf <==
+# Ansible managed
+
+#
+# Rules for the local system logs
+#
+ruleset(name="files") {
+  $IncludeConfig /etc/rsyslog.d/*.files
+}
+ruleset(name="forwards") {
+  $IncludeConfig /etc/rsyslog.d/*.forwards
+}
+ruleset(name="files_forwards") {
+  $IncludeConfig /etc/rsyslog.d/*.files
+  $IncludeConfig /etc/rsyslog.d/*.forwards
+}
+
+```
+In this PR, just files and forwards outputs are supported.  As described in [To Do's](#to-dos), elasticsearch is going to be added next.
+
+### Input roles
+Based upon the inventry file that users provide, input config jumps to the ruleset.  
+
+The next `imfile` example shows log messages read from /var/log/containers/*.log by the `imfile` plugin are logged in the local files as well as forwarded to the remote host.  
+
+**`imfile` example**
+```
+input(type="imfile" file="/var/log/containers/*.log" tag="container" ruleset="files_forwards")
+```
+The next `imjournal` example shows log messages from journald by the `imjournal` plugin are logged in the local files.
+
+**`imjournal` example**
+```
+module(load="imjournal"
+       StateFile="/var/lib/rsyslog/imjournal.state"
+       RateLimit.Burst="300000"
+       RateLimit.Interval="30"
+       PersistStateInterval="3000")
+call files
+```
+Please note that `imfile` input supports `ruleset` parameter, thus the parameter is used to specify which ruleset to call.
+Other plugins which do not have such a parameter, e.g., `imjournal`, use the `call` statement.
```

Comment:
The `call files` will be called for all inputs, not just the `imjournal` input - as opposed to the `input` with the `ruleset`.

Reply:
You could do something like this (not sure of the field names or the syntax):
```
if $inputname == "imjournal" then {
    call files
}
```

Reply:
In the rsyslog config, if it is ordered as follows, do you think `call files` is effective just for imjournal?  Or log messages from other im-plugins take the path?
```
module(load="imjournal" ...)
call files
input(name="basics_imtcp" type="imtcp ... ruleset="remote")
input(type="imfile" file="/path/to/input_file" ... ruleset="files_forwards")
```
The 2 input plugins im{tcp,udp} and imfile except imjournal support the input function, with which the ruleset to jump could be managed.

If the above order satisfies our purpose, I think we are good...

Reply:
I guess that would work - we would have to be very careful about the order of modules/inputs